modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+The type system chooses the correct function definition with the function dispatching algorithm.
+This returns null because of the definition of map for null.
+
+== Composite Types
+
+Composite types contains other values inside it. These composite types are Array, Object and Function.
+
+=== Array Type
+
+An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. +
+For example, the syntax to define an Array of Number as `Array&lt;Number&gt;` is:
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.
+
+=== Object Type
+
+In order to define an object type there are two ways, either `Object` or `{}`, they both define an _open object_ and does not specify any constraint to their key value pairs. Open and close objects are later defined in this section.
+
+The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:
+
+----
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields use `?`. For example, the following type allows the age field to be present or not:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only puts constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accept objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+----
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a close object is `{| |}`. The following example, makes our user type only supports the keys `firstName` and `lastName`, so this throws an exception because of the `age` field is not accepted.
```

Comment:
`The following example, makes our user type only support the keys `firstName` and `lastName`, so this throws an exception because the `